Add overview on what a package is (#170)
elastic/package-spec: README.md

In the following is a high level overview of a package.

## Asset organisation

In general, assets within a package are organised by {stack-component}/{asset-type}. For example assets for Elasticsearch ingest pipelines are in the folder `elasticsearch/ingest-pipeline`. The same logic applies to all Elasticsearch, Kibana and Elastic Agent assets.

There is a special folder `data_stream`. All assets inside the `data_stream` folder must follow the [Data Stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). `data_stream` can contain multiple folders, each with the name of that describes the content. Inside this folder, the same structure as before for {stack-component}/{asset-type} applies. The difference is that for all these assets, Fleet during installation enforces naming rules related to the [Data Stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). All assets in this folder belong directly or indirectly to data streams.

> In contrast, any asset added on the top level will be picked up as json document, pushed to the corresponding Elasticsearch / Kibana APIs and used as is. In most scenarios, only data stream assets are needed. There are exceptions where global assets are required to get more flexibility. This could be for example an ILM policy that applies to all data streams.

Reviewer: Thanks for adding the ILM example. Just copy editing:

```suggestion
In contrast, any asset added on the top level will be picked up as json document, pushed to the corresponding Elasticsearch / Kibana APIs and used as is. In most scenarios, only data stream assets are needed. There are exceptions where global assets are required to get more flexibility. This could be, for example, an ILM policy that applies to all data streams.
```

## Version Compatibility between Packages and Specifications

A package specifying its `format_version` as `x.y.z` must be valid against specifications in the semantic version range `[x.y.z, X.0.0)`, where `X = x + 1`.

# What is an Elastic Package?

> An Elastic Package is a collection of assets for the Elastic Stack. In addition, it contains manifest files which contain additional information about the package. The exact content and structure of a package are described by the package spec. In the following is a high level overview of a package.

Reviewer: Maybe add something about how packages also contain information about configuring them via the Fleet UI?

Author: Added

> A package is downloaded as a .zip file for the package-registry by Fleet inside Kibana. The zip file is then unpacked and each asset is installed into the related API.

Reviewer: ```suggestion
A package is downloaded as a .zip file from the package-registry by Fleet inside Kibana. The zip file is then unpacked and each asset is installed into the related API.
```

Reviewer: I think we can leave out the `as a .zip file` here. Feels like an implementation detail.

Author: Hm, not sure. I often explain a package as a glorified zip file.

Reviewer: What I mean is that the fact that a package is "downloaded as a .zip file" is entirely incidental as far as a package author or package consumer (end-user) is concerned. If I'm either of these personas, why should I care how a package is encoded in transport between the package registry and Kibana? Also tomorrow if we decide to deliver packages as `.tar` files or something else, are we going to remember to update this location here?

## Asset organisation

> In general, the assets are organised by {stack-component}/{asset-type}. For example assets for Elasticsearch ingest pipelines are in the directory `elasticsearch/ingest-pipeline`. The same logic applies to all Elasticsearch, Kibana and Elastic Agent assets.

Reviewer: s/directory/folder/ for consistency with other docs in this repo.

```suggestion
In general, assets within a package are organised by {stack-component}/{asset-type}. For example assets for Elasticsearch ingest pipelines are in the folder `elasticsearch/ingest-pipeline`. The same logic applies to all Elasticsearch, Kibana and Elastic Agent assets.
```

> There is a special directory `data_stream`. All assets inside the data_stream directory must follow the [Data Stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). `data_stream` can contain multiple directory, each with the name of the dataset (or postfix). Inside this directory, the same structure as before for {stack-component}/{asset-type} applies. The difference is that for all these assets, Fleet during installation enforces naming rules related to the [Data Stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). All assets in this directory belong directly or indirectly to data streams.

Reviewer: ```suggestion
There is a special folder `data_stream`. All assets inside the `data_stream` folder must follow the [Data Stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). `data_stream` can contain multiple folders, each with the name of the dataset (or postfix). Inside this folder, the same structure as before for {stack-component}/{asset-type} applies. The difference is that for all these assets, Fleet during installation enforces naming rules related to the [Data Stream naming scheme](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme). All assets in this folder belong directly or indirectly to data streams.
```

Reviewer: What is a dataset? 😄 

If we are going to introduce new terminology (relative to the rest of the documentation here) and especially if that terminology sounds similar but isn't quite the same as other terminology (dataset vs. data stream), we need to explain what this new terminology is.

Otherwise it's going to setup readers for confusion. Some will assume dataset is the same as data stream. Others will just be left confused and start questioning their understanding of data streams, etc.

Author: I linked to the data stream naming scheme that should explain it. But nobody is going to jump there to read up on it first. At the same time I do not want to dive to deep into this overview. Any ideas for a way out here?

Author: Modified it slightly to leave out dataset.

> In contrast, any asset added on the top level will be picked up as json document, pushed to the corresponding Elasticsearch / Kibana API and used as is. In most scenarios, only the data_stream assets are need, there are exceptions where global assets are required to get more flexibility.

Reviewer: ```suggestion
In contrast, any asset added on the top level will be picked up as json document, pushed to the corresponding Elasticsearch / Kibana APIs and used as is. In most scenarios, only data stream assets are needed. There are exceptions where global assets are required to get more flexibility.
```

Reviewer: Last sentence seems somewhat vague and not very actionable. Maybe give some examples of assets for that last sentence? Or examples of the type of flexibility?

If I'm a reader of the package spec documentation, I'm probably here to understand what goes where so I can do the right thing when building my package. Having concrete examples wil help with this.

Author: Added an ILM policy example.

## Supported assets

> In the following each currently supported asset type is listed for a quick overview. The package spec will always contain the fully up-to-date list:

Reviewer: ```suggestion
For a quick overview, these are the assets typically found in an Elastic Package. The package spec will always contain the fully up-to-date list.
```

* Elasticsearch
  * Ingest Pipeline
  * Index Template
  * Transform
  * Index template settings
* Kibana
  * Dashboards
  * Visualization
  * Index patterns
  * ML Modules
  * Map
  * Search
  * Security rules
* Other
  * fields.yml

> The special asset `fields.yml` is used to generate out of a single definition Elasticsearch Index Templates and Kibana index patterns. The exact definition can be found in the package spec.

Reviewer: ```suggestion
The special asset `fields.yml` is used to generate Elasticsearch Index Templates and Kibana index patterns. The exact definition can be found in the package spec.
```
